### PR TITLE
[MI-3287] Added API to get connected channels

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**

### CODE REVIEW

File: server/api.go [hunk 1]
```diff
@@ -332,6 +334,53 @@ func (a *API) disconnect(w http.ResponseWriter, r *http.Request) {
 	_, _ = w.Write([]byte("Your account has been disconnected."))
 }
 
+func (a *API) getConnectedChannels(w http.ResponseWriter, r *http.Request) {
+	if r.Method == http.MethodOptions {
+		return
+	}
+
+	userID := r.Header.Get("Mattermost-User-ID")
```
Comment (Kshitij-Katiyar) [inline server/api.go:342]: constant

File: server/api.go [hunk 2]
```diff
@@ -332,6 +334,53 @@ func (a *API) disconnect(w http.ResponseWriter, r *http.Request) {
...
+				MSTeamsChannelID:      link.MSTeamsChannelID,
+				MattermostChannelName: link.MattermostChannelName,
+				MattermostTeamName:    link.MattermostTeamName,
+				MSTeamsChannelName:    msTeamsChannelIDsVsNames[link.MSTeamsChannelID],
+				MSTeamsTeamName:       msTeamsTeamIDsVsNames[link.MSTeamsTeamID],
+			})
+		}
+	}
+
+	data, _ := json.Marshal(paginatedLinks)
```
Comment (Kshitij-Katiyar) [inline server/api.go:378]: check error

File: server/api.go [hunk 3]
```diff
@@ -332,6 +334,53 @@ func (a *API) disconnect(w http.ResponseWriter, r *http.Request) {
...
+
+	w.Header().Set("Content-Type", "application/json")
+	_, _ = w.Write(data)
+}
+
 // TODO: Add unit tests
 func (a *API) oauthRedirectHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (Kshitij-Katiyar) [inline server/api.go:385]: check error

File: server/api_test.go [hunk 4]
```diff
@@ -823,3 +825,113 @@ func TestDisconnect(t *testing.T) {
...
+	}{
+		{
+			Name: "GetConnectedChannels: linked channels listed successfully",
+			SetupPlugin: func(api *plugintest.API) {
+				api.On("HasPermissionToChannel", testutils.GetUserID(), testutils.GetChannelID(), model.PermissionCreatePost).Return(true).Times(1)
+			},
+			SetupStore: func(store *storemocks.Store) {
+				store.On("ListChannelLinksWithNames").Return([]*storemodels.ChannelLink{
+					{
+						MattermostTeamID:    testutils.GetTeamID(),
```
Comment (Kshitij-Katiyar) [inline server/api_test.go:848]: create util that will direct return ChannelLink

File: server/api_test.go [hunk 5]
```diff
@@ -823,3 +825,113 @@ func TestDisconnect(t *testing.T) {
...
+			plugin.configuration.EnforceConnectedUsers = test.EnforceConnectedUsers
+			plugin.configuration.EnabledTeams = test.EnabledTeams
+			test.SetupPlugin(plugin.API.(*plugintest.API))
+			test.SetupStore(plugin.store.(*storemocks.Store))
+			test.SetupClient(plugin.msteamsAppClient.(*clientmocks.Client))
+			w := httptest.NewRecorder()
+			r := httptest.NewRequest(http.MethodGet, "/get-connected-channels", nil)
+			r.Header.Add("Mattermost-User-ID", testutils.GetUserID())
+			queryParams := url.Values{
+				"per_page": {"10"},
```
Comment (Kshitij-Katiyar) [inline server/api_test.go:921]: constant for all of them

File: server/api_test.go [hunk 6]
```diff
@@ -823,3 +825,113 @@ func TestDisconnect(t *testing.T) {
...
+		t.Run(test.Name, func(t *testing.T) {
+			assert := assert.New(t)
+			plugin := newTestPlugin(t)
+			plugin.configuration.EnforceConnectedUsers = test.EnforceConnectedUsers
+			plugin.configuration.EnabledTeams = test.EnabledTeams
+			test.SetupPlugin(plugin.API.(*plugintest.API))
+			test.SetupStore(plugin.store.(*storemocks.Store))
+			test.SetupClient(plugin.msteamsAppClient.(*clientmocks.Client))
+			w := httptest.NewRecorder()
+			r := httptest.NewRequest(http.MethodGet, "/get-connected-channels", nil)
```
Comment (Kshitij-Katiyar) [inline server/api_test.go:918]: check whether there is a constant for this url

File: server/api_test.go [hunk 7]
```diff
@@ -823,3 +825,113 @@ func TestDisconnect(t *testing.T) {
...
+				"page":     {"0"},
+			}
+
+			r.URL.RawQuery = queryParams.Encode()
+			plugin.ServeHTTP(nil, w, r)
+			result := w.Result()
+			assert.NotNil(t, result)
+			defer result.Body.Close()
+
+			bodyBytes, _ := io.ReadAll(result.Body)
```
Comment (Kshitij-Katiyar) [inline server/api_test.go:931]: assert error

File: server/getTeamsData.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	"strconv"
+	"strings"
+
+	"github.com/mattermost/mattermost-plugin-msteams-sync/server/store/storemodels"
+	"github.com/mattermost/mattermost-server/v6/model"
+)
+
+const (
+	// Query params
+	QueryParamPerPage = "per_page"
```
Comment (Kshitij-Katiyar) [inline server/getTeamsData.go:14]: use them in test cases

File: server/api_test.go (hunk 6 above)
Comment (raghavaggarwal2308) [reply server/api_test.go:916]: No

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/api.go [hunk 9]
```diff
@@ -34,15 +36,16 @@ func NewAPI(p *Plugin, store store.Store) *API {
...
-	router.HandleFunc("/avatar/{userId:.*}", api.getAvatar).Methods(http.MethodGet)
-	router.HandleFunc("/changes", api.processActivity).Methods(http.MethodPost)
-	router.HandleFunc("/lifecycle", api.processLifecycle).Methods(http.MethodPost)
-	router.HandleFunc("/autocomplete/teams", api.autocompleteTeams).Methods(http.MethodGet)
-	router.HandleFunc("/autocomplete/channels", api.autocompleteChannels).Methods(http.MethodGet)
-	router.HandleFunc("/needsConnect", api.needsConnect).Methods(http.MethodGet, http.MethodOptions)
-	router.HandleFunc("/connect", api.connect).Methods(http.MethodGet, http.MethodOptions)
-	router.HandleFunc("/disconnect", api.handleAuthRequired(api.disconnect)).Methods(http.MethodGet, http.MethodOptions)
-	router.HandleFunc("/oauth-redirect", api.oauthRedirectHandler).Methods(http.MethodGet, http.MethodOptions)
+	router.HandleFunc("/avatar/{userId:.*}", api.getAvatar).Methods("GET")
```
Comment (ayusht2810) [inline server/api.go:39]: revert all changes to use HTTP method constants

File: server/api.go [hunk 10]
```diff
@@ -341,6 +344,64 @@ func (a *API) disconnect(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		a.p.API.LogError("Error occurred while getting the linked channels", "Error", err.Error())
+		http.Error(w, "Error occurred while getting the linked channels", http.StatusInternalServerError)
+		return
+	}
+
+	paginatedLinks := []*storemodels.ChannelLink{}
+	if len(links) > 0 {
+		sort.Slice(links, func(i, j int) bool {
+			return links[i].MattermostChannelID < links[j].MattermostChannelID
```
Comment (ayusht2810) [inline server/api.go:363]: any reason why we are sorting the links based on these conditions?

File: server/api.go [hunk 11]
```diff
@@ -341,6 +344,64 @@ func (a *API) disconnect(w http.ResponseWriter, r *http.Request) {
...
+
+		msTeamsTeamIDsVsNames, msTeamsChannelIDsVsNames, err := a.p.GetMSTeamsTeamAndChannelDetailsFromChannelLinks(links, userID, true)
+		if err != nil {
+			a.p.API.LogError("Unable to get the MS Teams channels and teams detail", "Error", err.Error())
+			http.Error(w, "Unable to get the MS Teams channels and teams detail", http.StatusInternalServerError)
+			return
+		}
+
+		offset, limit := a.p.GetOffsetAndLimitFromQueryParams(r)
+		for index, link := range links {
```
Comment (ayusht2810) [inline server/api.go:374]: can we directly start our loop from the offset index to avoid extra iterations and computations?

File: server/api.go [hunk 12]
```diff
@@ -341,6 +344,64 @@ func (a *API) disconnect(w http.ResponseWriter, r *http.Request) {
...
+			if len(paginatedLinks) == limit {
+				break
+			}
+
+			channel, appErr := a.p.API.GetChannel(link.MattermostChannelID)
+			if appErr != nil {
+				a.p.API.LogError("Error occurred while getting the channel details", "ChannelID", link.MattermostChannelID, "Error", appErr.Message)
+				http.Error(w, "Error occurred while getting the channel details", http.StatusInternalServerError)
+				return
```
Comment (ayusht2810) [inline server/api.go:383]: I think we should use continue here to move to the next link

File: server/api.go [hunk 13]
```diff
@@ -341,6 +344,64 @@ func (a *API) disconnect(w http.ResponseWriter, r *http.Request) {
...
+		offset, limit := a.p.GetOffsetAndLimitFromQueryParams(r)
+		for index, link := range links {
+			if len(paginatedLinks) == limit {
+				break
+			}
+
+			channel, appErr := a.p.API.GetChannel(link.MattermostChannelID)
+			if appErr != nil {
+				a.p.API.LogError("Error occurred while getting the channel details", "ChannelID", link.MattermostChannelID, "Error", appErr.Message)
+				http.Error(w, "Error occurred while getting the channel details", http.StatusInternalServerError)
```
Comment (ayusht2810) [inline server/api.go:382]:
```suggestion
```

File: server/api_test.go [hunk 14]
```diff
@@ -829,3 +831,139 @@ func TestDisconnect(t *testing.T) {
...
+			SetupStore: func(store *storemocks.Store) {
+				store.On("ListChannelLinksWithNames").Return([]*storemodels.ChannelLink{
+					testutils.GetChannelLink(),
+				}, nil).Times(1)
+			},
+			SetupClient: func(c *clientmocks.Client) {
+				c.On("GetChannelsInTeam", testutils.GetTeamID(), fmt.Sprintf("id in ('%s')", testutils.GetChannelID())).Return([]*msteams.Channel{}, nil).Times(1)
+				c.On("GetTeams", fmt.Sprintf("id in ('%s')", testutils.GetTeamID())).Return(nil, errors.New("error occurred while getting the team details")).Times(1)
+			},
+			ExpectedResult:     "Unable to get the MS Teams channels and teams detail\n",
```
Comment (ayusht2810) [inline server/api_test.go:900]:
```suggestion
			ExpectedResult:     "Unable to get the MS Teams teams details\n",
```

File: server/api_test.go [hunk 15]
```diff
@@ -829,3 +831,139 @@ func TestDisconnect(t *testing.T) {
...
+			},
+			SetupStore: func(store *storemocks.Store) {
+				store.On("ListChannelLinksWithNames").Return(nil, errors.New("error occurred while getting the linked channels")).Times(1)
+			},
+			SetupClient:        func(c *clientmocks.Client) {},
+			ExpectedResult:     "Error occurred while getting the linked channels\n",
+			ExpectedStatusCode: http.StatusInternalServerError,
+		},
+		{
+			Name: "GetConnectedChannels: error occurred while getting the team details",
```
Comment (ayusht2810) [inline server/api_test.go:885]: same

File: server/api_test.go [hunk 16]
```diff
@@ -829,3 +831,139 @@ func TestDisconnect(t *testing.T) {
...
+			},
+			SetupClient: func(c *clientmocks.Client) {
+				c.On("GetChannelsInTeam", testutils.GetTeamID(), fmt.Sprintf("id in ('%s')", testutils.GetChannelID())).Return([]*msteams.Channel{}, nil).Times(1)
+				c.On("GetTeams", fmt.Sprintf("id in ('%s')", testutils.GetTeamID())).Return(nil, errors.New("error occurred while getting the team details")).Times(1)
+			},
+			ExpectedResult:     "Unable to get the MS Teams channels and teams detail\n",
+			ExpectedStatusCode: http.StatusInternalServerError,
+		},
+		{
+			Name: "GetConnectedChannels: error occurred while getting channel details",
```
Comment (ayusht2810) [inline server/api_test.go:904]: same

File: server/getTeamsData.go [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+)
+
+func (p *Plugin) GetMSTeamsTeamAndChannelDetailsFromChannelLinks(channelLinks []*storemodels.ChannelLink, userID string, checkChannelPermissions bool) (map[string]string, map[string]string, error) {
+	msTeamsTeamIDsVsNames := make(map[string]string)
+	msTeamsChannelIDsVsNames := make(map[string]string)
+	msTeamsTeamIDsVsChannelsQuery := make(map[string]string)
+
+	for _, link := range channelLinks {
+		if checkChannelPermissions && !p.API.HasPermissionToChannel(userID, link.MattermostChannelID, model.PermissionCreatePost) {
+			p.API.LogInfo("User does not have the permissions for the requested channel", "UserID", userID, "ChannelID", link.MattermostChannelID)
```
Comment (ayusht2810) [inline server/getTeamsData.go:30]: do we really need this?

File: server/getTeamsData.go [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+
+	return nil
+}
+
+func (p *Plugin) GetMSTeamsChannelDetailsForAllTeams(msTeamsTeamIDsVsChannelsQuery, msTeamsChannelIDsVsNames map[string]string) error {
+	for teamID, channelsQuery := range msTeamsTeamIDsVsChannelsQuery {
+		channels, err := p.msteamsAppClient.GetChannelsInTeam(teamID, channelsQuery+")")
+		if err != nil {
+			p.API.LogDebug("Unable to get the MS Teams channel information for the team", "TeamID", teamID, "Error", err.Error())
+			return err
```
Comment (ayusht2810) [inline server/getTeamsData.go:88]: should we return or continue, as there can be more data that we can fetch?

File: server/api.go (hunk 12 above)
Comment (raghavaggarwal2308) [reply server/api.go:383]: You won't get the channel type for some channels in such case.Will that be fine?

File: server/getTeamsData.go (hunk 17 above)
Comment (raghavaggarwal2308) [reply server/getTeamsData.go:30]: Are you talking about the log statement of the check?

File: server/api.go (hunk 12 above)
Comment (raghavaggarwal2308) [reply server/api.go:383]: discussed

File: server/api.go [hunk 19]
```diff
@@ -358,30 +358,26 @@ func (a *API) getConnectedChannels(w http.ResponseWriter, r *http.Request) {
...
-			channel, appErr := a.p.API.GetChannel(link.MattermostChannelID)
-			if appErr != nil {
-				a.p.API.LogError("Error occurred while getting the channel details", "ChannelID", link.MattermostChannelID, "Error", appErr.Message)
-				http.Error(w, "Error occurred while getting the channel details", http.StatusInternalServerError)
-				return
-			}
-
+		for index := offset; index < offset+limit && index < len(links); index++ {
+			link := links[index]
 			if index >= offset && msTeamsChannelIDsVsNames[link.MSTeamsChannelID] != "" && msTeamsTeamIDsVsNames[link.MSTeamsTeamID] != "" {
```
Comment (ayusht2810) [inline server/api.go:373]:
```suggestion
			if msTeamsChannelIDsVsNames[link.MSTeamsChannelID] != "" && msTeamsTeamIDsVsNames[link.MSTeamsTeamID] != "" {
```

File: server/getTeamsData.go [hunk 20]
```diff
@@ -44,21 +44,19 @@ func (p *Plugin) GetMSTeamsTeamAndChannelDetailsFromChannelLinks(channelLinks []
...
 	}
 
-	var resultingErr error
+	errorsFound := false
 	// Get MS Teams display names for each unique team ID and store it
-	if err := p.GetMSTeamsTeamDetails(msTeamsTeamIDsVsNames); err != nil {
-		resultingErr = err
-	}
+	teamDetailsErr := p.GetMSTeamsTeamDetails(msTeamsTeamIDsVsNames)
+	errorsFound = errorsFound || teamDetailsErr
```
Comment (ayusht2810) [inline server/getTeamsData.go:50]:
```suggestion
	if err := p.GetMSTeamsTeamDetails(msTeamsTeamIDsVsNames); err != nil {
	    errorsFound = true
	}
```

File: server/getTeamsData.go [hunk 21]
```diff
@@ -44,21 +44,19 @@ func (p *Plugin) GetMSTeamsTeamAndChannelDetailsFromChannelLinks(channelLinks []
...
 
 	// Get MS Teams channel details for all channels for each unique team
-	if err := p.GetMSTeamsChannelDetailsForAllTeams(msTeamsTeamIDsVsChannelsQuery, msTeamsChannelIDsVsNames); err != nil {
-		resultingErr = err
-	}
+	channelDetailsErr := p.GetMSTeamsChannelDetailsForAllTeams(msTeamsTeamIDsVsChannelsQuery, msTeamsChannelIDsVsNames)
+	errorsFound = errorsFound || channelDetailsErr
```
Comment (ayusht2810) [inline server/getTeamsData.go:54]: same

File: server/getTeamsData.go [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+		}
+
+		msTeamsTeamIDsVsChannelsQuery[link.MSTeamsTeamID] += "'" + link.MSTeamsChannelID + "'"
+	}
+
+	errorsFound := false
+	// Get MS Teams display names for each unique team ID and store it
+	if p.GetMSTeamsTeamDetails(msTeamsTeamIDsVsNames) {
+		errorsFound = true
+	}
```
Comment (manojmalik20) [inline server/getTeamsData.go:51]:
```suggestion
	// Get MS Teams display names for each unique team ID and store it
	errorsFound := p.GetMSTeamsTeamDetails(msTeamsTeamIDsVsNames)
```

File: server/getTeamsData.go [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+	QueryParamPerPage = "per_page"
+	QueryParamPage    = "page"
+
+	// Pagination
+	DefaultPage         = 0
+	DefaultPerPageLimit = 10
+	MaxPerPageLimit     = 100
+)
+
+func (p *Plugin) GetMSTeamsTeamAndChannelDetailsFromChannelLinks(channelLinks []*storemodels.ChannelLink, userID string, checkChannelPermissions bool) (map[string]string, map[string]string, bool) {
```
Comment (manojmalik20) [inline server/getTeamsData.go:23]: Give a name to the return variables or add documentation above the function.

File: server/getTeamsData.go [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+
+	// Get MS Teams channel details for all channels for each unique team
+	if p.GetMSTeamsChannelDetailsForAllTeams(msTeamsTeamIDsVsChannelsQuery, msTeamsChannelIDsVsNames) {
+		errorsFound = true
+	}
+
+	return msTeamsTeamIDsVsNames, msTeamsChannelIDsVsNames, errorsFound
+}
+
+func (p *Plugin) GetMSTeamsTeamDetails(msTeamsTeamIDsVsNames map[string]string) bool {
```
Comment (manojmalik20) [inline server/getTeamsData.go:61]: Give a name to the return variable or add documentation above the function.

File: server/getTeamsData.go [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+
+		for _, channel := range channels {
+			msTeamsChannelIDsVsNames[channel.ID] = channel.DisplayName
+		}
+	}
+
+	return errorsFound
+}
+
+func (p *Plugin) GetOffsetAndLimitFromQueryParams(r *http.Request) (offset, limit int) {
```
Comment (manojmalik20) [inline server/getTeamsData.go:101]:
```suggestion
func (p *Plugin) GetOffsetAndLimit(r *http.Request) (offset, limit int) {
```

File: server/api_test.go [hunk 26]
```diff
@@ -829,3 +831,138 @@ func TestDisconnect(t *testing.T) {
...
+			test.SetupPlugin(mockAPI)
+			test.SetupStore(plugin.store.(*storemocks.Store))
+			test.SetupClient(plugin.msteamsAppClient.(*clientmocks.Client))
+
+			w := httptest.NewRecorder()
+			r := httptest.NewRequest(http.MethodGet, "/connected-channels", nil)
+			r.Header.Add(HeaderMattermostUserID, testutils.GetUserID())
+			queryParams := url.Values{
+				QueryParamPerPage: {fmt.Sprintf("%d", DefaultPerPageLimit)},
+				QueryParamPage:    {fmt.Sprintf("%d", DefaultPage)},
```
Comment (manojmalik20) [inline server/api_test.go:951]:
```suggestion
				QueryParamPerPage: {fmt.Sprint(DefaultPerPageLimit)},
				QueryParamPage:    {fmt.Sprint(DefaultPage)},
```

**Pull request merged**